File: src/main/java/com/lambdaschool/starthere/services/StoryServiceImpl.java
package com.lambdaschool.starthere.services;

import com.lambdaschool.starthere.exceptions.ResourceNotFoundException;
import com.lambdaschool.starthere.models.Story;
import com.lambdaschool.starthere.repository.StoryRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;

@Service(value = "storyService")
public class StoryServiceImpl implements StoryService
{
    @Autowired
    private StoryRepository storyRepository;

    @Override
    public List<Story> findAll()
    {
        List<Story> list = new ArrayList<>();
        storyRepository.findAll().iterator().forEachRemaining(list::add);
        return list;
    }

    @Override
    public Story findQuoteById(long id)
    {
        return storyRepository.findById(id).orElseThrow(() -> new ResourceNotFoundException(Long.toString(id)));
    }

    @Override
    public void delete(long id)
    {
        if (storyRepository.findById(id).isPresent())
        {
            Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
            if (storyRepository.findById(id).get().getUser().getUsername().equalsIgnoreCase(authentication.getName()))
            {
                storyRepository.deleteById(id);
            } else
            {
                throw new ResourceNotFoundException(id + " " + authentication.getName());
            }
        } else
        {
            throw new ResourceNotFoundException(Long.toString(id));
        }
    }

    @Transactional
    @Override
    public Story save(Story story)
    {
        return storyRepository.save(story);
    }

    @Override
    public List<Story> findByUserName(String username)
    {
        List<Story> list = new ArrayList<>();
        storyRepository.findAll().iterator().forEachRemaining(list::add);

        list.removeIf(q -> !q.getUser().getUsername().equalsIgnoreCase(username));
        return list;
    }
    @Transactional
    @Override
    public Story update(Story story, Long id)
    {
        Story currentStory = storyRepository.findById(id).orElseThrow();
        if (story.getTitle() != null){
            currentStory.setTitle(story.getTitle());
        }
        if (story.getCountry() != null){
            currentStory.setCountry(story.getCountry());
        }
        if (story.getDescription() != null){
            currentStory.setDescription(story.getDescription());
        }
        if (story.getContent() != null){
            currentStory.setContent(story.getContent());
        }
        if (story.getDate() != null){
            currentStory.setDate(story.getDate());
        }
        if (story.getUser() != null){
            currentStory.setUser(story.getUser());
        }

        return storyRepository.save(story);
    }


}

File: src/main/kotlin/com/example/kotlinstory/services/StoryServiceImpl.kt
package com.example.kotlinstory.services

import com.example.kotlinstory.exceptions.ResourceNotFoundException
import com.example.kotlinstory.models.Story
import com.example.kotlinstory.repository.StoryRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

import javax.persistence.EntityNotFoundException
import java.util.ArrayList
import java.util.function.Consumer
@Service(value = "storyService")
open class StoryServiceImpl : StoryService {
    @Autowired
    private val storyrepo: StoryRepository? = null

    override fun findAll(): List<Story> {
        val list = ArrayList<Story>()
        storyrepo!!.findAll().iterator().forEachRemaining(Consumer<Story> { list.add(it) })
        return list
    }

    override fun findStoryById(id: Long): Story {
        return storyrepo!!.findById(id)
                .orElseThrow { EntityNotFoundException(id.toString()) }
    }

    override fun delete(id: Long) {
        if (storyrepo!!.findById(id).isPresent) {
            val authentication = SecurityContextHolder.getContext().authentication
            if (storyrepo.findById(id).get().user!!.username!!.equals(authentication.name, ignoreCase = true)) {
                storyrepo.deleteById(id)
            } else {
                throw EntityNotFoundException(id.toString() + " " + authentication.name)
            }
        } else {
            throw EntityNotFoundException(id.toString())
        }
    }

    @Transactional
    override fun save(quote: Story): Story {
        return storyrepo!!.save(quote)
    }

    override fun findByUserName(username: String): List<Story> {
        val list = ArrayList<Story>()
        storyrepo!!.findAll().iterator().forEachRemaining(Consumer<Story> { list.add(it) })

        list.removeIf { q -> !q.user!!.username!!.equals(username, ignoreCase = true) }
        return list
    }

    override fun update(id: Long, book: Story) {
        val currentBook = storyrepo!!.findById(id)
                .orElseThrow { ResourceNotFoundException("Cannot find book id$id") }

        if (book.title != null) {
            currentBook.title = book.title
        }

        if (book.country != null) {
            currentBook.country = book.country
        }

        if (book.description != null) {
            currentBook.description = book.description
        }
        //todo will need to change if date is different type (other than string)
        if (book.date != null) {
            currentBook.date = book.date
        }


        storyrepo!!.save(currentBook)
    }
}
